### PR TITLE
Renderable logic improvements to web viewer

### PR DESCRIPTION
**Pull request by kwokcb**
# Changes

- Fixes: #1633
- Adds in proper parsing of renderable elements vs always just picking the first surface shader found.
  - The test suite files which have (multiple) nodegraph outputs and top level outputs will now load properly.
- Adds in UI to mark folders as renderable (using a shaderball icon) 
- Adds in "soloing" capability to allow picking a renderable and have it show up on all geometry. When nothing is solo'ed the default material assignment is used.
- Includes fixes for:
  - Dag path assignment matching .
  - Missing exposure of NodeGraph::getDownStreamPorts() in JS.
  - Addressing the big performance hit when binding materials to geometry in ThreeJS. The code by default is some quite slow code for reflection / debugging purposes which is now turned off. Chess set load is seconds vs minutes. This affects the 'solo'ing workflow significantly as each switch is a geometry re-bind. If the material is not already cached then slow code will be hit which can cause seconds to pass when selecting a new material -- which appears like a "hang" from a user perspective.

# Results

## Videos
- Shows basic interaction of picking different renderables to solo and editing inputs on solo'ed item.

https://github.com/AcademySoftwareFoundation/MaterialX/assets/49369885/908536a3-4a3b-4b3d-99b3-4cad1fb2507b

- Shows improved bind time performance for 2 cases: 
  - Load and bind all programs on document load (entire chess set)
  - Load all programs but incrementally binding on renderable switch
Once cached the program bind time overhead is reduced.

https://github.com/AcademySoftwareFoundation/MaterialX/assets/49369885/e17324b4-4b26-4aa2-bc14-93eaead3e8ae

## Snapshots
- Includes case where are already assignments. Solo'ing is non-destructive and `materialassign` assigments found on load will be restored on unsolo'ing. (wood example)

| On load | Solo'ed |
| :--: | :--: | 
| image | image |
| image | image  image |
|  image |  image  image |

### CODE REVIEW

File: javascript/MaterialXView/source/viewer.js [hunk 1]
```diff
@@ -769,31 +926,108 @@ export class Material
...
+
+        // Create and cache associated UI
+        var matUI = gui.addFolder(elemPath);        
+        matassign.setMaterialUI(matUI);
+
+        let matTitle = matUI.domElement.getElementsByClassName('title')[0];
+        // Add a icon to the title to allow for assigning the material to geometry
+        // Clicking on the icon will "solo" the material to the geometry.
+        // Clicking on the title will open/close the material folder.
+        matTitle.innerHTML = "<img id='" + elemPath + "' src='public/shader_ball.svg' width='16' height='16' style='vertical-align:middle; margin-right: 5px;'>" + elem.getNamePath();
```
Comment (kwokcb) [inline javascript/MaterialXView/source/viewer.js:1016]: This mechanism of inserting additional DOM elements + event handler allows for insertion of additional interaction. We can do this anywhere we'd like to add in additional action "buttons".

File: javascript/MaterialXView/source/viewer.js [hunk 2]
```diff
@@ -769,31 +926,108 @@ export class Material
...
+                img.src = 'public/shader_ball.svg';
+                //matTitle.classList.remove('peditor_material_unassigned');
+            }
+        }
+    }
+
+    static updateSoloMaterial(viewer, elemPath, materials, event)
+    {
+        // Prevent the event from being passed to parent folder
+        event.stopPropagation();
```
Comment (kwokcb) [inline javascript/MaterialXView/source/viewer.js:957]: This is the only "trick" to capture the "icon" events to not being passed to the parent action which is to collapse the folder.

File: javascript/MaterialXView/source/viewer.js [hunk 3]
```diff
@@ -623,15 +667,110 @@ export class Material
...
+            {
+                let nodeGraph = nodeGraphs[i];
+                if (nodeGraph)
+                {
+                    if (nodeGraph.hasAttribute('nodedef') || nodeGraph.hasSourceUri())
+                    {
+                        continue;
+                    } 
+                    // Skip any nodegraph that is connected to something downstream
+                    if (nodeGraph.getDownstreamPorts().length > 0)
```
Comment (kwokcb) [inline javascript/MaterialXView/source/viewer.js:712]: This needs the missing API exposure. Prevents re-including nodegraphs which are already connected to downstream shaders.

**Review by jstone-lucasfilm — APPROVED** [review]
This looks good to me, thanks @kwokcb!

**Comment by hybridherbst** [comment]
Awesome! Thank you so much!

**Pull request merged**